google/cloud/spanner/internal/connection_impl_test.cc

```diff
@@ -2056,6 +2034,296 @@ TEST(ConnectionImplTest, TransactionOutlivesConnection) {
   conn.reset();
 }
 
+MATCHER(HasBadSession, "bound to a session that's marked bad") {
+  return internal::Visit(
+      arg, [&](internal::SessionHolder& session,
+               google::spanner::v1::TransactionSelector&, std::int64_t) {
+        if (!session) {
+          *result_listener << "has no session";
+          return false;
+        }
+        if (!session->is_bad()) {
+          *result_listener << "session expected to be bad, but was not";
+          return false;
+        }
+        return true;
+      });
+}
+
+TEST(ConnectionImplTest, Read_SessionNotFound) {
```

Comment:
nit: Test names are not supposed to include underscores (et seq).

Reply:
Right, I know. But that seems to be the form used in the rest of this file, so I was following consistency.

I believe the same is true in other test files as well. I think we should do a separate PR to clean this up if we want to.

```shell
TEST(ConnectionImplTest, Read_PermanentFailure) {
TEST(ConnectionImplTest, Read_TooManyTransientFailures) {
TEST(ConnectionImplTest, ExecuteDmlDelete_PermanentFailure) {
TEST(ConnectionImplTest, ExecuteDmlDelete_TooManyTransientFailures) {
TEST(ConnectionImplTest, ProfileDmlDelete_PermanentFailure) {
TEST(ConnectionImplTest, ProfileDmlDelete_TooManyTransientFailures) {
TEST(ConnectionImplTest, AnalyzeSqlDelete_PermanentFailure) {
TEST(ConnectionImplTest, AnalyzeSqlDelete_TooManyTransientFailures) {
TEST(ConnectionImplTest, ExecuteBatchDml_PartialFailure) {
TEST(ConnectionImplTest, ExecuteBatchDml_PermanmentFailure) {
TEST(ConnectionImplTest, ExecuteBatchDml_TooManyTransientFailures) {
TEST(ConnectionImplTest, ExecutePartitionedDmlDelete_PermanentFailure) {
TEST(ConnectionImplTest, ExecutePartitionedDmlDelete_TooManyTransientFailures) {
TEST(ConnectionImplTest, CommitGetSession_PermanentFailure) {
TEST(ConnectionImplTest, CommitGetSession_TooManyTransientFailures) {
TEST(ConnectionImplTest, CommitGetSession_Retry) {
TEST(ConnectionImplTest, CommitCommit_PermanentFailure) {
TEST(ConnectionImplTest, CommitCommit_TooManyTransientFailures) {
TEST(ConnectionImplTest, CommitCommit_IdempotentTransientSuccess) {
TEST(ConnectionImplTest, CommitCommit_NonIdempotentTransientFailure) {
TEST(ConnectionImplTest, Rollback_PermanentFailure) {
TEST(ConnectionImplTest, Rollback_TooManyTransientFailures) {
TEST(ConnectionImplTest, Rollback_Success) {
TEST(ConnectionImplTest, PartitionRead_PermanentFailure) {
TEST(ConnectionImplTest, PartitionRead_TooManyTransientFailures) {
TEST(ConnectionImplTest, PartitionQuery_PermanentFailure) {
TEST(ConnectionImplTest, PartitionQuery_TooManyTransientFailures) {
TEST(ConnectionImplTest, Read_SessionNotFound) {
TEST(ConnectionImplTest, PartitionRead_SessionNotFound) {
TEST(ConnectionImplTest, ExecuteQuery_SessionNotFound) {
TEST(ConnectionImplTest, ProfileQuery_SessionNotFound) {
TEST(ConnectionImplTest, ExecuteDml_SessionNotFound) {
TEST(ConnectionImplTest, ProfileDml_SessionNotFound) {
TEST(ConnectionImplTest, AnalyzeSql_SessionNotFound) {
TEST(ConnectionImplTest, PartitionQuery_SessionNotFound) {
TEST(ConnectionImplTest, ExecuteBatchDml_SessionNotFound) {
TEST(ConnectionImplTest, ExecutePartitionedDml_SessionNotFound) {
TEST(ConnectionImplTest, Commit_SessionNotFound) {
TEST(ConnectionImplTest, Rollback_SessionNotFound) {
```

Reply:
Filed https://github.com/googleapis/google-cloud-cpp-spanner/issues/1144 to track this.